### ISSUE TITLE
cmd/go: clarify go.mod documentation for trivial relative modules

### ISSUE THREAD

**Issue by gazed** [golang/go#27274]
### What version of Go are you using (`go version`)?
<pre><code>go version go1.11 darwin/amd64
</code></pre>

### Does this issue reproduce with the latest release?
Yes.

### What operating system and processor architecture are you using (`go env`)?
<pre><code>GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/rust/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/rust/code"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/rust/code/modtest/b/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/_d/8n_wrc4n32d1rgx80bn6hsl80000gn/T/go-build056371648=/tmp/go-build -gno-record-gcc-switches -fno-common"
</code></pre>

### What did you do?

Created two modules <code>a</code> and <code>b</code> in the same directory, not on the <code>GOPATH</code>.

Module <code>a</code> contains <code>a.go, go.mod</code>:
<pre><code>~/code/modtest/a: cat a.go
package a

func Hello() string {
	return "Hello"
}
~/code/modtest/a: cat go.mod
module a
</code></pre>

Module <code>b</code> contains <code>b.go, go.mod</code>:
<pre><code>~/code/modtest/b: cat b.go
package main

import (
    "fmt"

    "a"
)

func main() {
	fmt.Printf("%s\n", a.Hello())
}
~/code/modtest/b: cat go.mod 
module b

require ../a v0.0.0
</code></pre>

What is the proper syntax for the line <code>require ../a v0.0.0</code> in this particular case?
The closest I found in the docs (<code>go help modules, go help go.mod, go help mod</code>) was from <code>go help importpath</code>:
<pre><code>Relative import paths

An import path beginning with ./ or ../ is called a relative path.
The toolchain supports relative import paths as a shortcut in two ways.

First, a relative path can be used as a shorthand on the command line.
...

Second, if you are compiling a Go program not in a work space,
you can use a relative path in an import statement in that program
to refer to nearby code also not in a work space.
This makes it easy to experiment with small multipackage programs
outside of the usual work spaces, but such programs cannot be
installed with "go install" (there is no work space in which to install them),
so they are rebuilt from scratch each time they are built.
To avoid ambiguity, Go programs cannot use relative import paths
within a work space.
</code></pre>
It is unclear to me if the second case is supposed to apply in this situation.

### What did you expect to see?

I expect this to compile and run similarly to how it works with the existing GOPATH.
For example, copying the modules to the GOPATH gives the following:
<pre><code>~/code/src: cd a
~/code/src/a: go build
~/code/src/a: cd ../b
~/code/src/b: go build
~/code/src/b: ./b
Hello
</code></pre>

### What did you see instead?

<pre><code>~/code/modtest: cd a
~/code/modtest/a: go build
~/code/modtest/a: cd ../b
~/code/modtest/b: go build
go: ../a@v0.0.0: unrecognized import path "../a" (https fetch: Get https://../a?go-get=1: dial tcp: lookup ..: no such host)
go: error loading module requirements
~/code/modtest/b: 
</code></pre>

**Comment by thepudds** [comment]
Setting aside for the moment the question of whether the documentation could be improved, here is a concrete example that might help.

Sample file structure on my local system, all outside of GOPATH:
```
    /tmp/playground/hello
    |-- go.mod
    `-- hello.go
    /tmp/playground/goodbye
    |-- go.mod
    `-- goodbye.go
```

And the contents of  /tmp/playground/hello/go.mod:
``` 
    module example.com/me/hello

    require (
     example.com/me/goodbye v0.0.0
     rsc.io/quote v1.5.2
    )

    replace example.com/me/goodbye => ../goodbye
```

See a longer discussion and the rest of the example here (which is a runnable example if you are interested):
https://groups.google.com/d/msg/golang-nuts/1nYoAMFZVVM/eppaRW2rCAAJ

**Comment by thepudds** [comment]
@gopherbot please add label modules

**Comment by gazed** [comment]
Thank you for the working module solution. Setting aside for the moment whether or not this is the canonical solution moving forward, it breaks when moving the code to the GOPATH, making the code non-backwards compatible. Is there another solution that works without changing the source code files? Otherwise it feels like a missing feature or a bug, instead of "a me not finding the right documentation" problem as I had assumed.

Here is the original problem source updated to work with modules.
<pre></code>~/code/modtest/a: cat a.go
package a

func Hello() string {
	return "Hello"
}
~/code/modtest/a: cat go.mod 
module example.com/me/a
~/code/modtest/a: go build
~/code/modtest/a: cd ../b
~/code/modtest/b: cat b.go 
package main

import (
    "fmt"
    
    "example.com/me/a"
)

func main() {
	fmt.Printf("%s\n", a.Hello())
}
~/code/modtest/b: cat go.mod 
module example.com/me/b

require example.com/me/a v0.0.0

replace example.com/me/a => ../a
~/code/modtest/b: go build
~/code/modtest/b: ./b
Hello
</code></pre>

Moving both modules into the GOPATH results in the following:
<pre><code>~/code/src: cd a
~/code/src/a: go build
~/code/src/a: cd ../b
~/code/src/b: go build
b.go:6:2: cannot find package "example.com/me/a" in any of:
	/usr/local/go/src/example.com/me/a (from $GOROOT)
	/Users/rust/code/src/example.com/me/a (from $GOPATH)
</code></pre>

**Comment by thepudds** [comment]
If you return to the GOPATH world, then you would place the go source code back where Go 1.10 and earlier would expect it based on the import paths you are using in the code.

E.g., something like `$GOPATH/src/example.com/me/a/a.go`?

If you don't like that `example.com`, then you don't need to use it in your `go.mod` and corresponding import paths.

And sorry if I am misunderstanding; this is all new...

**Comment by gazed** [comment]
Thank you for your patience. It is more likely me either not understanding or not explaining the problem clearly. My understanding of your example is that it involves externally visible packages. My concern is for internal development. My use case is my internal game engine <code>a</code> with a number of sub-packages (created back in the Go 1.0 timeframe). I use the game engine <code>a</code> to create games, in this case modules <code>b,c,d,e...</code>.  The game engine <code>a</code> works fine with modules since there are no external dependencies. However none of the games I create (eg <code>b,c,d,e...</code>) are currently working with modules because they can't locate the game engine <code>a</code>. Currently all the modules <code>a,b,c,d,...</code> are in the same directory on the GOPATH.

My example using <code>a,b</code> modules was to simplify the problem for the purposes of reporting functionality that used to be working and no longer works with Go 1.11 modules. If you are suggesting I need to change my code base in order to work with modules, then I most certainly will - even though I will probably wait until the Go1.12 timeframe when modules become mandatory. However I was hoping (had assumed) that the problem was in the go.mod "require" line. It just felt to me like it should be able to handle relative modules as GOPATH does now.

Am I wrong to think that modules, as it stands now, breaks backwards compatibility? Or, more likely, the internal development scenario described above was never intended to be supported by the Go tools? -- mainly because I'm not bothering with the company based prefixes since I'm not publishing my code externally?

To be fair, I do publish a public version of my engine to <code>https://github.com/gazed/vu</code> This is done once or twice a year with a small python script that modifies the internal package paths when pushing to github for external consumption. However I find my development cycle much easier working with the internal view of the engine as I am constantly updating and refining it. As well I have about 10 game projects at this point that refer to the engine as <code>vu</code> and not <code>github.com/gazed/vu</code>. The other benefit I find with this solution is that I can move my externally visible engine code to a different repository (it may happen) and my internal game and game engine imports don't have to change.

**Comment by klaus** [comment]
should be documented as `go mod edit -replace a=../a`

want to add that for me, this does NOT work with a module that's deeper in the dependency (`go mod edit -replace a/b=../a/b`). The go mod does work, but I still get 

my go.mod (of mycomp.com/hauskey-enhancer):
replace mycomp.com/post-adressdaten/addressdata => ../post-adressdaten/addressdata

go build:
go build mycomp.com/hauskey-enhancer: cannot find module for path mycomp.com/post-adressdaten/addressdata

**Comment by klaus** [comment]
it's important to note, that you need the "require" AND the "replace" in the go.mod file.
I like to add them via the commandline, the commands should be encouraged and documented with examples.

the `go mod init` command does not populate go.mod with packages it cannot resolve. Therefore you must do:

go mod edit require example.com/post-adressdaten
go mod edit -replace example.com/post-adressdaten=../post-adressdaten/

**Comment by myitcv** [comment]
@klaus 

> I like to add them via the commandline, the commands should be encouraged and documented with examples.

They are documented under `go help mod edit`

That a `replace` requires (!) a `require` is something that is going to be addressed in Go 1.12 via https://github.com/golang/go/issues/26241

I'm going to close this issue on the basis I think all questions have been answered, but please shout it not.

**Comment by dfang** [comment]
@myitcv how about this case ?

```
/tmp/modtest1X9IT λ  tree
.
├── cmd
│   └── main.go
├── go.mod
└── handler.go

1 directory, 3 files
/tmp/modtest1X9IT λ  cat go.mod
module modtest

go 1.12
/tmp/modtest1X9IT λ  cat handler.go
package handler

import (
    "fmt"
)

func Handler() {
	fmt.Println("hello")
}
/tmp/modtest1X9IT λ  cat cmd/main.go
package main

import "handler"

func main() {
	// how to call handler.Handler here
	handler.Handler()
}
```

how to call handler.Handler in cmd/main.go if code is not hosted on github but on cloud storage or private google cloud source repo ?

I'd like to run cmd/main.go like this [codelab](https://codelabs.developers.google.com/codelabs/cloud-functions-go-http/index.html?index=..%2F..io2019#5) for testing handler locally

thanks !
